8300891: Deprecate for removal javax.swing.plaf.synth.SynthLookAndFeel.load(URL url)

prrace opened the pull request:
SynthLookAndFeel.load(URL) is problematic because applications need to trust the URL.
As documented in the bug report there are alternative mechanisms and JDK itself has not used this for the two Synth based L&Fs : Nimbus and GTK.
So deprecating - for removal - and adding warnings now in the docs is appropriate.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8301032](https://bugs.openjdk.org/browse/JDK-8301032) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8300891](https://bugs.openjdk.org/browse/JDK-8300891): Deprecate for removal javax.swing.plaf.synth.SynthLookAndFeel.load(URL url)
 * [JDK-8301032](https://bugs.openjdk.org/browse/JDK-8301032): Deprecate for removal javax.swing.plaf.synth.SynthLookAndFeel.load(URL url) (**CSR**)


### Reviewers
 * [Sergey Bylokhov](https://openjdk.org/census#serb) (@mrserb - **Reviewer**) ⚠️ Review applies to [738414c0](https://git.openjdk.org/jdk/pull/12175/files/738414c0fb21060ad7afdb6df1abefd9366849f7)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Alexander Zuev](https://openjdk.org/census#kizune) (@azuev-java - **Reviewer**)
 * [Prasanta Sadhukhan](https://openjdk.org/census#psadhukhan) (@prsadhuk - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12175/head:pull/12175` \
`$ git checkout pull/12175`

Update a local copy of the PR: \
`$ git checkout pull/12175` \
`$ git pull https://git.openjdk.org/jdk pull/12175/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12175`

View PR using the GUI difftool: \
`$ git pr show -t 12175`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12175.diff">https://git.openjdk.org/jdk/pull/12175.diff</a>

</details>

prrace commented:
/csr

This needs a CSR because of the deprecation

mrserb commented:
I do not think that this method should/must be used by custom L&F only, it is a way to provide custom styles for an existing L&f including Nimbus like discussed here:
https://stackoverflow.com/questions/11564597/how-to-use-synthlookandfeel-with-xml-file-where-xml-file-path-will-load-the-fil
There is a way to check the permission - security manager, ohh we deprecated it as well, time to rethink that? or do we plan to delete all methods which use URL, or add custom properties here and there to check access?

prrace commented:
> I do not think that this method should/must be used by custom L&F only, it is a way to provide custom styles for an existing L&f including Nimbus like discussed here:
> https://stackoverflow.com/questions/11564597/how-to-use-synthlookandfeel-with-xml-file-where-xml-file-path-will-load-the-fil
> There is a way to check the permission - security manager, ohh we deprecated it as well, time to rethink that? or do we plan to delete all methods which use URL, or add custom properties here and there to check access?

Who knows, we may deprecate more methods when there's a good reason,
but this one regardless of whether it is the sole source of styles for a Synth L&F or customisation is a prime candidate for deprecation and subsequent removal.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLookAndFeel.java:
```diff
@@ -615,6 +615,16 @@ public void load(InputStream input, Class<?> resourceBase) throws
+     * equal level of trust with application code, using it to load from
+     * from remote resources, particularly any which may have a lower level of
+     * trust, is strongly discouraged.
+     * The alternative mechanisms to load styles from an {@code InputStream}
+     * located as a resource co-located with the application or by
```
“…an `InputStream` *located* as a resource co-located…”

Will it sound clearer if the word “located” is removed?

prrace replied:
"using a resource co-located with the application"  ?

I'm going to write up the CSR now since the wording seems mostly settled.

aivanov-jdk replied:
Yes, it looks better to me. Thank you.

prrace commented:
The CSR https://bugs.openjdk.org/browse/JDK-8301032 is now ready for a reviewer.
I did not include the doc-files changes since I don't think the comments there are required in the CSR.

prsadhuk commented on src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLookAndFeel.java:
```diff
@@ -615,6 +615,16 @@ public void load(InputStream input, Class<?> resourceBase) throws
+     * The alternative mechanisms to load styles from an {@code InputStream}
+     * using a resource co-located with the application or by
+     * providing a {@code SynthStyleFactory} are preferred.
+     * Consequently this method is deprecated and will be removed in a future release.
```
We probably can use "@see" tag to link the alternate preferred method `load(InputStream input, Class<?> resourceBase)` since this is going to be deprecated

aivanov-jdk replied:
Great suggestion. The javadoc should link to the alternatives `load` or `SynthStyleFactory`.

The following line should be added to javadoc:
```java
     * @deprecated Use {@link #load(InputStream, Class)} or {@link SynthStyleFactory}
     *             as described above.
```
The entire paragraph could probably used in `@deprecated`. The result would look like this:
![Rendered javadoc with @deprecated before the added paragraph](https://user-images.githubusercontent.com/70774172/215103931-2f093a8a-f057-49f7-8dce-e8cbb9e2ff90.png)

Here's the code that I used:
```java
     * information.
     *
     * @deprecated Whilst this API may be safe for loading local resources that are
     * ...
     * The alternative mechanisms to load styles from an
     * {@link #load(InputStream, Class) InputStream} using a resource
     * co-located with the application or by
     * providing a {@link SynthStyleFactory} are preferred.
     * ...
```
However, such use of linking for `load` with `InputStream` is likely to be confusing, so the text should be re-worded.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLookAndFeel.java:
```diff
@@ -615,6 +615,16 @@ public void load(InputStream input, Class<?> resourceBase) throws
+     * <p>
+     * Whilst this API may be safe for loading local resources that are
+     * delivered with a {@code LookAndFeel} or application, and so have an
+     * equal level of trust with application code, using it to load from
+     * from remote resources, particularly any which may have a lower level of
```
```suggestion
     * equal level of trust with application code, using it to load from
     * remote resources, particularly any which may have a lower level of
```
The word _“from”_ is repeated, thank to IDEA for catching it.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLookAndFeel.java:
```diff
@@ -615,6 +615,16 @@ public void load(InputStream input, Class<?> resourceBase) throws
+     * The alternative mechanisms to load styles from an {@code InputStream}
+     * using a resource co-located with the application or by
+     * providing a {@code SynthStyleFactory} are preferred.
+     * Consequently this method is deprecated and will be removed in a future release.
```
```suggestion
     * Consequently, this method is deprecated and will be removed in a future release.
```
Shall the comma be added?

prrace commented:
I added the links and removed the from but the "," and moving the text doesn't work well as
the @deprecated tag is placed first by the javadoc and I think it was better to go into the explanation where it was.
Also "as described above" doesn't work because there's nothing above ..

prsadhuk commented on src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLookAndFeel.java:
```diff
@@ -615,14 +615,28 @@ public void load(InputStream input, Class<?> resourceBase) throws
      * @throws IOException if synthSet cannot be opened as an <code>InputStream</code>
      * @since 1.6
+     * @deprecated Use {@link load(InputStream, Class)} or
+     * {@link setStyleFactory} instead
```
The link to setStyleFactory was used like "{@link #setStyleFactory}" earlier
https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLookAndFeel.java#L77

shouldn't we keep it same?

prrace replied:
where was this ? And why ? This works just fine.

prsadhuk replied:
I guess there's different ways but AFAIS, java.desktop javadocs use 
"@link xxx to link to a class
and "@link #xxx" to link to a field, constant or method in a class

for example, in SYnthLookAndFeel class only, we have this 
_{@link #load} method, or provide your own {@link SynthStyleFactory}
  to {@link #setStyleFactory}_

ANother example 
```
java/awt/Color.java:     * Throws an {@link IllegalArgumentException} if the value is out of
java/awt/Color.java:     * {@link ColorModel}.
java/awt/Color.java:     * format used by the method {@link #getRGB() getRGB}.
```

and here setStyleFactory is a method, so I asked...

prrace commented:
It doesn't seem to matter .. but I added # .. and I changed one of the cases to use linkplain - the one embedded in the text - because it looks better that way. Cleaned up some formatting also.

prrace commented:
/integrate
